File: demos/test-cpp/CodingInterviewChinese2/s0002_add_two_numbers.cpp
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

class Solution {
   public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int total = 0;
        ListNode* head = new ListNode(-1);
        ListNode* t = head;
        while (l1 != NULL || l2 != NULL || total > 0) {
            if (l1 != NULL) {
                total += l1->val;
                l1 = l1->next;
            }
            if (l2 != NULL) {
                total += l2->val;
                l2 = l2->next;
            }
            if (total >= 10) {
                total -= 10;
                carry = 1;
            }
            t->next = new ListNode(total);
            t = t->next;

            // 忘记设置会死循环
            total = carry;
            carry = 0;
        }
        return head->next;
    }
};
int main() {
    return 0;
}

File: demos/test-rust/coding_interview/src/solution/s0002_add_two_numbers.rs
#![allow(unused_imports)]
#![allow(dead_code)]

use crate::base_st::ListNode;

pub struct Solution {}

impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut l1 = l1;
        let mut l2 = l2;
        let mut carry = 0;
        let mut total = 0;
        let mut r = ListNode::new(0);
        let mut head = &mut r;

        while l1.is_some() || l2.is_some() || total > 0 {
            if l1.is_some() {
                total = total + l1.as_ref()?.val;
                l1 = l1?.next;
            }
            if l2.is_some() {
                total = total + l2.as_ref()?.val;
                l2 = l2?.next;
            }
            if total >= 10 {
                total -= 10;
                carry = 1;
            }
            head.next = Some(Box::new(ListNode {
                val: total,
                next: None,
            }));
            head = head.next.as_mut()?;

            total = carry;
            carry = 0;
        }

        r.next
    }
}
